File: week3/MyRCoursework/code/get_TreeHeight.py
import csv
import sys
import os
import math

# Function to calculate tree height using trigonometry
def calculate_tree_height(degrees, distance):
    # Convert angle from degrees to radians for trigonometric calculation
    radians = math.radians(degrees)
    # Calculate height using the tangent of the angle and distance
    return distance * math.tan(radians)

# Function to process data from a CSV file
def process(csv_path):    
    # Lists to store processed data and headers
    data = []
    headers = []

    # Open and read the CSV file
    with open(csv_path, 'r') as csv_file:
        csv_reader = csv.reader(csv_file)
        # Extract the header row
        headers = next(csv_reader)

        # Iterate through each row in the CSV file
        for row in csv_reader:
            # Identify column indices for angle and distance
            angle_idx = headers.index("Angle.degrees")
            distance_idx = headers.index("Distance.m")

            # Convert angle and distance values to floats
            angle_degrees = float(row[angle_idx])
            distance = float(row[distance_idx])

            # Calculate tree height and append to the row
            tree_height = calculate_tree_height(angle_degrees, distance)
            row.append(tree_height)
            data.append(row)

    # Append the new header for tree height
    headers.append("Tree.Height.m")
    return headers, data

# Function to write data to a new CSV file
def save_to_csv(savepath, headers, data):    
    # Open the output file for writing
    with open(savepath, 'w', newline='') as csv_file:
        csv_writer = csv.writer(csv_file)
        # Write the header row
        csv_writer.writerow(headers)
        # Write each data row
        for row in data:
            csv_writer.writerow(row)

# Main function to orchestrate the processing of the CSV file
def main(argv):   
    # Check for the presence of command-line argument
    if len(sys.argv) < 2:
        print("Usage: get_TreeHeight.py <filepath of csv>")
        return 1  # Error code for incorrect usage

    # Retrieve the path of the CSV file from command-line argument
    csv_path = sys.argv[1]
    # Process the CSV file
    headers, data = process(csv_path)
    # Construct a file path for saving the results
    base_name = os.path.splitext(os.path.basename(csv_path))[0]
    save_path = os.path.join("../results", f"{base_name}_treeheights.csv")
    # Save the processed data to the new CSV file
    save_to_csv(save_path, headers, data)

    return 0  # Success code

# Script entry point
if (__name__ == "__main__"):
    status = main(sys.argv)
    sys.exit(status)

File: week3/MyRCoursework/code/get_TreeHeight.R
# Define the function to calculate tree height
TreeHeight <- function(degrees, distance) {
    radians <- degrees * pi / 180
    height <- distance * tan(radians)
    return (height)
}

# Obtain the command line arguments
args <- commandArgs(trailingOnly = TRUE)

# Ensure a file is provided as an argument
if(length(args) == 0) {
    stop("No file provided. Usage: get_TreeHeight.R <filename.csv>")
}

input_file <- args[1]

# Load the data
data <- read.csv(input_file, header=TRUE)

# Calculate tree heights
data$Tree.Height.m <- mapply(TreeHeight, data$Angle.degrees, data$Distance.m)

# Extract the base file name without extension and path
base_name <- tools::file_path_sans_ext(basename(input_file))

# Create the output file name
output_file <- paste0("../results/", base_name, "_treeheights.csv")

# Write the combined data to the output file
write.csv(data, file=output_file, row.names=FALSE)
